hydra/hydra.go

```diff
@@ -0,0 +1,541 @@
+// Package hydra implements the hydra user consent flow
+package hydra
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/ashtonian/authboss/hydra/hconsenter"
+	"github.com/volatiletech/authboss"
+	"golang.org/x/crypto/bcrypt"
+)
+
+type ConsentValuer interface {
+	authboss.Validator
+
+	GetScopes() []string
+	GetIsAllowed() bool
+	GetRequestedAudience() []string
+}
+
+func MustHaveConsent(v authboss.Validator) ConsentValuer {
+	if u, ok := v.(ConsentValuer); ok {
+		return u
+	}
+
+	panic(fmt.Sprintf("bodyreader returned a type that could not be upgraded to ConsentValuer: %T", v))
+}
+
+type ChallengeValuer interface {
+	authboss.Validator
+
+	GetChallenge() string
+}
+
+func MustHaveChallenge(v authboss.Validator) ChallengeValuer {
+	if u, ok := v.(ChallengeValuer); ok {
+		return u
+	}
+
+	panic(fmt.Sprintf("bodyreader returned a type that could not be upgraded to ChallengeValuer: %T", v))
+}
+
+type LogoutValuer interface {
+	authboss.Validator
+
+	GetShouldLogout() bool
+}
+
+func MustHaveLougout(v authboss.Validator) LogoutValuer {
+	if u, ok := v.(LogoutValuer); ok {
+		return u
+	}
+
+	panic(fmt.Sprintf("bodyreader returned a type that could not be upgraded to LogoutValuer: %T", v))
+}
+
+type SessionableUser interface {
+	authboss.User
+
+	GetSession() (session map[string]interface{})
+}
+
+// TODO: document oauth2 and openID reserved scopes/session keys, potentially type them via an additional module
+const (
+	// PageLogin is for identifying the login page for parsing & validation
+	PageLogin    = "login"
+	PageConsent  = "consent"
+	PageLogout   = "logout"
+	ChallengeKey = "challenge"
+)
+
+func init() {
+	authboss.RegisterModule("hydraconsent", &HydraConsent{})
+}
+
+// HydraConsent module
+type HydraConsent struct {
+	*authboss.Authboss
+	hClient                   *hconsenter.Client
+	rememberMe                int
+	ignoreConsent             map[string]bool
+	overrideRequestedAudience bool
+}
+
+// Init module
+func (a *HydraConsent) Init(ab *authboss.Authboss) (err error) {
+	a.Authboss = ab
+	a.rememberMe = 3600
+
+	if err = a.Authboss.Config.Core.ViewRenderer.Load(PageLogin); err != nil {
```

Comment:
These can be a single line:
https://github.com/volatiletech/authboss/blob/master/renderer.go#L8